### DISCUSSION TITLE
Source Decomposition Options for `PermissionSets`

### DISCUSSION THREAD

**Discussion by VivekMChawla** [forcedotcom/cli#2993] [Polls]
## The design for decomposing `PermissionSets` has been finalized.

Thanks to everyone who voted and shared their thoughts regarding the design of a **new source behavior option** for `PermissionSet` metadata. Based on your feedback, we will proceed with **Option Two: Focused Decomposition**.

The [PermissionSet Source Decomposition](https://salesforce.vidyard.com/watch/Dw5v2UhrXvnxGaX5mxFvmK?) video remains available for anyone who'd like to see examples of "Focused Decomposition." 

You can also review the details of the [Focused Decomposition design](https://github.com/forcedotcom/cli/discussions/2993#discussioncomment-10431670) (Option Two), below.

**Comment by VivekMChawla** [comment]
## 📺 Watch [PermissionSet Source Decomposition](https://salesforce.vidyard.com/watch/Dw5v2UhrXvnxGaX5mxFvmK?) to review the proposed options.

![CLI_Permset_Vote_GitHub](https://github.com/user-attachments/assets/023a1313-1aab-4a8d-80cd-744540a220b0)

# [↑ Cast Your Vote ↑](https://github.com/forcedotcom/cli/discussions/2993#discussion-7087365)

**Comment by VivekMChawla** [comment]
# Option One: Total Decomposition

Remember, all new source behaviors are **optional**. Your source files won't change unless you activate this feature.

## Proposed Source Behavior

* Each `PermissionSet` component is decomposed into an individual folder.
    * Each folder must contain a single `permissionset-meta.xml` source file.
    * The `permissionset-meta.xml` source file name must match the containing folder name.
    * Each `permissionset-meta.xml` source file must be unique inside a project. 
* All permissions and settings are decomposed into individual files and grouped into category folders.
    * Each file is named for the individual permission or setting it represents.
    * ⚠️ A single, large permission set may result in **hundreds** of decomposed source files.
* Permissions and settings for objects are spread across multiple files, which are spread across multiple folders.

## Example

```
sfdx-project
 └─ force-app
     └─ main
         └─ default
             └─ permissionsets
                 ├─ PO_Manager
                 │   ├─ applicationVisibilities 
                 │   │   └─ PURCHASE.applicationVisibility-meta.xml 
                 │   ├─ classAccesses 
                 │   │   ├─ PurchaseOrderController.classAccess-meta.xml 
                 │   │   ├─ PurchaseOrderTransfer.classAccess-meta.xml 
                 │   │   └─ VendorListController.classAccess-meta.xml 
                 │   ├─ customPermissions
                 │   │   ├─ Purchase_Manager.customPermission-meta.xml 
                 │   │   └─ Purchase_User.customPermission-meta.xml 
                 │   ├─ customSettingAccesses
                 │   │   ├─ PurchaseApp_Manager_Settings__c.customSettingAccess-meta.xml 
                 │   │   └─ PurchaseApp_User_Settings__c.customSettingAccess-meta.xml 
                 │   ├─ externalCredentialPrincipalAccesses
                 │   │   └─ Quote_Verification_Service.externalCredentialPrincipalAccess-meta.xml 
                 │   ├─ externalDataSourceAccesses
                 │   │   ├─ Approved_Vendors__x.externalDataSourceAccess-meta.xml 
                 │   │   └─ Job_Request__c.objectSettings-meta.xml 
                 │   ├─ fieldPermissions
                 │   │   ├─ Account.Vendor_Account_ID__c.fieldPermission-meta.xml 
                 │   │   ├─ Account.Vendor_Status__c.fieldPermission-meta.xml 
                 │   │   ├─ PO_Line_Item__c.Net_Cost__c.fieldPermission-meta.xml 
                 │   │   ├─ PO_Line_Item__c.Total__c.fieldPermission-meta.xml 
                 │   │   ├─ Purchase_Order__c.Approval_Date__c.fieldPermission-meta.xml 
                 │   │   ├─ Purchase_Order__c.Invoice_Number__c.fieldPermission-meta.xml 
                 │   │   ├─ Purchase_Order__c.Purchasing_Manager__c.fieldPermission-meta.xml 
                 │   │   ├─ Purchase_Order__c.Total__c.fieldPermission-meta.xml 
                 │   │   ├─ Purchase_Order__c.Tracking_Numbers__c.fieldPermission-meta.xml 
                 │   │   └─ Purchase_Order__c.Vendor_Contact__c.fieldPermission-meta.xml 
                 │   ├─ flowAccesses
                 │   │   ├─ Approve_Purchase_Order.flowAccess-meta.xml 
                 │   │   └─ Create_Purchase_Order.flowAccess-meta.xml 
                 │   ├─ objectPermissions 
                 │   │   ├─ Account.objectPermission-meta.xml 
                 │   │   ├─ PO_Line_Item__c.objectPermission-meta.xml 
                 │   │   └─ Purchase_Order__c.objectPermission-meta.xml 
                 │   ├─ pageAccesses 
                 │   │   ├─ PurchaseOrder.pageAccess-meta.xml 
                 │   │   └─ VendorList.pageAccess-meta.xml 
                 │   ├─ recordTypeVisibilities
                 │   │   ├─ Purchase_Order__c.Closed_PO.recordTypeVisibility-meta.xml 
                 │   │   └─ Purchase_Order__c.Open_PO.recordTypeVisibility-meta.xml 
                 │   ├─ tabSettings
                 │   │   ├─ Purchase_Order__c.tabSetting-meta.xml 
                 │   │   └─ Purchase_Order_Summary.tabSetting-meta.xml 
                 │   ├─ userPermissions
                 │   │   ├─ DataExport.userPermission-meta.xml 
                 │   │   ├─ DelegatedTwoFactor.userPermission-meta.xml 
                 │   │   ├─ ViewAllCustomSettings.userPermission-meta.xml 
                 │   │   ├─ ViewEncryptedData.userPermission-meta.xml 
                 │   │   ├─ ViewPublicDashboards.userPermission-meta.xml 
                 │   │   └─ ViewPublicReports.userPermission-meta.xml 
                 │   └─ PO_Manager.permissionset-meta.xml
                 │
                 └─ PO_User // Folder closed for simplicity.
```

# [↑ Cast Your Vote ↑](https://github.com/forcedotcom/cli/discussions/2993#discussion-7087365)

**Comment by VivekMChawla** [comment]
# Option Two: Focused Decomposition

Remember, all new source behaviors are **optional**. Your source files won't change unless you activate this feature.

## Proposed Source Behavior

* Each `PermissionSet` component is decomposed into an individual folder.
    * Each folder must contain a single `permissionset-meta.xml` source file.
    * The `permissionset-meta.xml` source file name must match the containing folder name.
* Permissions and settings are decomposed into focused files, grouped by category.
* Focused permission/setting files use the Developer Name of the PermissionSet and a category-specific extension, for example:
    * `MyPermSet.applicationVisibilities-meta.xml` 
    * `MyPermSet.classAccesses-meta.xml` 
    * `MyPermSet.userPermissions-meta.xml`
* "Object Settings" consolidate four categories of perms/settings into a single `Object_Name.objectSettings-meta.xml` file.
    * `objectPermissions`
    * `fieldPermissions`
    * `tabSettings`
    * `recordTypeVisibilities`

## Example 

```
sfdx-project
 └─ force-app
     └─ main
         └─ default
             └─ permissionsets
                 ├─ PO_Manager
                 │   ├─ objectSettings 
                 │   │   ├─ Account.objectSettings-meta.xml 
                 │   │   ├─ PO_Line_Item__c.objectSettings-meta.xml 
                 │   │   └─ Purchase_Order__c.objectSettings-meta.xml 
                 │   ├─ PO_Manager.applicationVisibilities-meta.xml 
                 │   ├─ PO_Manager.classAccesses-meta.xml 
                 │   ├─ PO_Manager.customPermissions-meta.xml
                 │   ├─ PO_Manager.customSettingAccesses-meta.xml 
                 │   ├─ PO_Manager.externalCredentialPrincipalAccesses-meta.xml 
                 │   ├─ PO_Manager.externalDataSourceAccesses-meta.xml  
                 │   ├─ PO_Manager.flowAccesses-meta.xml 
                 │   ├─ PO_Manager.pageAccesses-meta.xml 
                 │   ├─ PO_Manager.permissionset-meta.xml
                 │   └─ PO_Manager.userPermissions-meta.xml 
                 │
                 └─ PO_User // Folder closed for simplicity.
```

# [↑ Cast Your Vote ↑](https://github.com/forcedotcom/cli/discussions/2993#discussion-7087365)

**Comment by codefriar** [comment]
@VivekMChawla  - I beg of you, whichever method you move forward with. Please make the order of elements in the xml deterministic. ie: please sort them... I just did a code review on a permission set. Developer added perms to a field. But because the retrieve command writes the permset xml ... in any order it sees fit - comparing it against a git version can be quite the chore.

**Reply by renatoliveira** [reply]
I think that's the case only for option 2, because in the Total Decomposition option the metadata files do not contain lists in general, so ordering does not matter at all considering each piece is a file on its own (and therefore ordering is up to your file browser).

This is actually one of the reasons I voted for option 1. Seems easier to inspect and maintain in the long run.

**Reply by VivekMChawla** [reply]
@codefriar - I agree with you. Sorted XML elements make sense. We should be able to do this, but I'll engage with the engineering team to confirm.

@renatoliveira - Could you elaborate on why individual files seem easier to inspect and maintain in the long run?

**Reply by renatoliveira** [reply]
@VivekMChawla I just find it easier to view and edit permissions if they are decomposed into smaller files. Having them in individual files also makes it easier to author scripts to modify/clean things up if needed in automated pipelines for larger projects, I think.

This is not an issue for new and smaller projects, but for big implementations the length of those XML files sometimes surprise me. I'm thinking that having them decomposed into individual files might help with cloning/editing without accessing the org's UI - for example: if someone adds some permissions to the wrong permission set by chance, it is easier to just move the files than edit the XML. I suspect this could then be done even by someone without much knowledge or experience with the source files and probably source control in general.

**Reply by jongpie** [reply]
I completely agree with @renatoliveira, having individual files ([option 1](https://github.com/forcedotcom/cli/discussions/2993#discussioncomment-10431664)) feels easier to view & edit, compared to big XML files. And I don't have any data to back it up, but in my own experience, I feel like having separate smaller files (using total decomposition) reduces the likelihood of having merge conflicts in `git`, compared to larger files (using either no decomposition like the old MDAPI format, or even focused decomposition).

Option 1 is also consistent with how custom objects are currently structured, where there are separate files for every custom field (instead of 1 big XML file containing all fields), list views, validation rules, etc. For example:

```
sfdx-project
 └─ force-app
     └─ main
         └─ default
             └─ objects
                 └─ Some_Object__c
                     ├─ fields
                     │   └─ Some_Field__c.field-meta.xml 
                     ├─ listViews
                     │   └─ All.listView-meta.xml
                     └─ validationRules
                         └─ Some_Validation_Rule.validationRule-meta.xml 
                
```

I maintain [Nebula Logger](https://github.com/jongpie/NebulaLogger/), which has a dozens of fields on some of its objects (hundreds of fields throughout the whole project), and the custom object structure being totally decomposed has been an amazing improvement for me when managing the project's metadata (compared to the old MDAPI format of 1 giant XML file for the whole object). My worry with focused decomposition is that for sizable projects, the XML files will still be big & hard to manage, even if they're more focused/smaller than the old format.

**Reply by jongpie** [reply]
One other big benefit with total decomposition: individual metadata items can be excluded from deployments/retrievals using `.forceignore`. Focused decomposition won't support this.
- With custom objects today (total decomposition), I can exclude a single field (or multiple fields) from being deployed by adding them to `.forceignore`.
- If all the fields were stored in a single file (focused decomposition), excluding _some_ fields would be a much bigger challenge for developers/pipelines/automation. The XML nodes to exclude would have to be stripped out of the XML file, and `.forceignore` wouldn't be able to help.

**Reply by VivekMChawla** [reply]
@jongpie - Thank you very much for your feedback.  

I want to clarify a few things about how `PermissionSet` metadata retrieval/deployment works.

1. Individual settings/perms within a permission set are not deployable.
2. This differs from types like custom objects that do have sub-components that are atomically deployable.
3. If a portion of a decomposed `PermissionSet` were _somehow_ excluded from deployment, it would result in modification of that `PermissionSet`.

I'm glad you brought up `.forceignore`.  I'll check with the engineering team to ensure protection against settings that would unintentionally result in destructive changes. 

By the way, I highly recommend reading this [extraordinary post](https://github.com/forcedotcom/cli/discussions/2544#discussioncomment-9045716) by @scolladon. It describes the three "types" of possible decomposability with Salesforce Metadata. It helped me wrap my head around this topic when I started digging into it.

**Reply by jongpie** [reply]
@VivekMChawla thanks for the clarification! And for linking to that post by @scolladon. It has some incredibly helpful info, and I wasn't aware of the official terms `ListableDecomposed`, `StandaloneDecomposed`, and `ContainedDecomposed`. And thanks @scolladon for putting together that post!

**Reply by sirephil** [reply]
I was slightly concerned over additional complexity in managing adding and deleting individual files in version control as perm sets are updated, which is why I voted for the focused decomposition, but I do see the possible advantages for full decomposition outlined above.

From a version control perspective, I prefer to keep supporting scripting and manual activities to a minimum which is another reason to go with focused decomposition.

Deterministic structuring of the XML itself is very important.

**Reply by horzel-evh** [reply]
> @VivekMChawla thanks for the clarification! And for linking to that post by @scolladon. It has some incredibly helpful info, and I wasn't aware of the official terms `ListableDecomposed`, `StandaloneDecomposed`, and `ContainedDecomposed`. And thanks @scolladon for putting together that post!

Reading following reply:
https://github.com/forcedotcom/cli/discussions/2544#discussioncomment-9049531
Three mentioned terms do not seem to be official. But it is a great post.

**Reply by VivekMChawla** [reply]
@jongpie - As noted by @horzel those terms aren't exactly "official". They're just a really good way to describe the different options.  Tip of my cap to @scolladon for sure! 😄

**Comment by rbklaassen** [comment]
Total decomposition will result in a very big amount of small files for each permission set. Imagine that you need to review a pull request with quite some permission sets :)
I like the **focused decomposition**, that would really be an improvement coming from 'composed' permission sets!

**Comment by pgonzaleznetwork** [comment]
@VivekMChawla  thanks a lot for working on this. Dumb question, why do we need to choose? Both options can satisfy different use cases. For example, when comparing diffs for a single change, having a single file to point to would be a dream, and it can also reduce merge conflicts. For comparing overall diffs across 2 perm sets, focused decomposition would help as well.

**Reply by VivekMChawla** [reply]
@pgonzaleznetwork - Thanks for asking this question.

Behind the scenes, the CLI does a lot of complex work when converting metadata between MDAPI and SFDX source formats. We want to offer new decomposition options, but each one increases the chance of introducing unexpected issues.

Focusing on a single decomposition method per metadata type reduces the surface area for troubleshooting and support. It's our way of balancing the introduction of new capabilities with our commitment to maintaining the stability the CLI.

**Reply by mcarvin8** [reply]
@pgonzaleznetwork  Feel free to use my plugin if you want total decomposition of perm sets. https://github.com/mcarvin8/sf-decomposer

If your perm sets are large, this becomes a pain since it will create hundreds/thousands of files. 

But this is an option for manageable perms if you want to have files for each perm.

**Comment by abd3** [comment]
@VivekMChawla - implicit in the idea that "Each permissionset-meta.xml source file must be unique inside a project." is the idea that a permission set can only exist inside one subfolder in the project. That is different from the way that Fields and Object metadata can be spread across multiple packages. I do think that's a better design, the potential for confusion is even greater if the permission sets are defined across multiple folders.

Total decomposition seems like a disaster. You can see all the names of the fields, but you can't see the permissions unless you open the file. The benefit of defining each CustomField is that they can be spread across multiple packages. THat's not relevant here. Total decomposition seems shortsighted.

**Reply by VivekMChawla** [reply]
> Implicit in the idea that "Each `permissionset-meta.xml` source file must be unique inside a project." is the idea that a permission set can only exist inside one subfolder in the project. 

@abd3 - YES! You understood my meaning perfectly! 😄

**Comment by rickroesler** [comment]
@VivekMChawla 
Focused decomposition. 
- you don't have to deal with hundreds of tiny files
- consistent with the UI

Some Elements.cloud-specific questions/feedback, but probably applies to any metadata-scanning product
- echoing @codefriar: PLEASE sort; creating change logs for these is difficult when the response is not determistic
- is this only for the CLI? or will this option be exposed via the metadata retrieve() api?
- if retrieve() api ... which would be awesome ... is it feature-flag-selectable (so that it won't break current integrations)
- PermissionSet only? One of the problems we have is with very large orgs with very large Profiles that break the batch heap size when we try to process them; this would be a huge help with that

**Reply by VivekMChawla** [reply]
@rickroesler - Thank you for engaging and providing feedback!

* +1 to the idea of sorting the XML nodes inside each file.  We're actively exploring ways to do this for each new source behavior option.
* The new source behavior options transform MDAPI files to SFDX source files. Nothing changes with the MDAPI.
* The new source behavior options are enabled at the SFDX project level, not the command or ENV level.  Multiple projects on the same machine can use different combinations of source behavior options without interfering with one another.
* Profiles are being considered but have particular complexities that make them difficult to decompose.
   * Could you elaborate on Profiles breaking the batch heap size when processing? 
   * Are you referring to processing the XML of a profile inside of Apex and not being able to do so under 12MB?  
   * If so, these source decomposition options won't help because they're exclusive to the local environment of an SFDX project. 

PS: If you're at DF this year, I hope we run into each other! 😄

**Reply by rickroesler** [reply]
@VivekMChawla 

- Yes to "processing the XML of a profile inside of Apex and not being able to do so under 12MB?" We have a couple of customers who are maxing out their objects and fields and they're all on the Sys Admin profile. We're moving that processing to AWS, but I thought this could help.
- I'm at DF. Presenting at two sessions: "Mastering Data Cloud: Optimal Implementation Strategies".

**Comment by joelopy** [comment]
Some thoughts on dev experience.

Total decomposition reduces the chance for merge conflicts when adding/removing mdt, which is nice.

However, keyword pollution on field names, etc. could increase friction during development.

Many small files, while providing granular control, can result in a confusing experience. Similar perm sets are defined unique via the path, not filename. We would have to be much more careful to ensure the file being edited is correct by comparing paths, and using the command palette (VS Code) to open files will offer a much larger list of options due to the new naming conventions.

**Comment by VivekMChawla** [comment]
Thank you all for voting and commenting!

Based on your feedback, we will proceed with **Option Two: Focused Decomposition** for `PermissionSet` metadata.

I'll now close and unpin this discussion. Thank you!

**Comment by mcarvin8** [comment]
@VivekMChawla - I'm not sure if this has been reported by anyone else using the newly released `decomposePermissionSetBeta2`, but I'm seeing strange behavior when decomposing all permission sets in my org with it, specifically with `*.customPermission-meta.xml` files when decomposed.

All decomposed `*.customPermission-meta.xml` files have no element name in between the tags as seen in screenshot, so these will not work since these aren't valid XMLs:

![{E2727F3B-FF42-4C37-99B1-C96E0A1F039D}](https://github.com/user-attachments/assets/afc938ca-dbaa-4c92-a6ef-ac5cdb5bbaeb)

I'm not seeing anything off in https://github.com/forcedotcom/source-deploy-retrieve/blob/main/src/registry/presets/decomposePermissionSetBeta2.json besides the `"isAddressable": false,`, which will just make these decomposed permission sets unable to be retrieved and deployed individually without the parent. Which is something I'm trying to fix with the workflow preset and sharing actions preset via https://github.com/forcedotcom/source-deploy-retrieve/pull/1467

That's a different issue, but decomposed items should be addressable (deployable/retrievable) without the parent in my opinion.

I don't know if anyone else has the same issue with `*.customPermission-meta.xml` files in their org if they are using this preset, but we have a rather large org.

**Reply by VivekMChawla** [reply]
@mcarvin8 - Thanks for highlighting this issue.  

Would you mind creating a new issue for this, including the screenshot and your notes?  I'll work with @WillieRuemmele to prioritize a fix once he's back from holiday PTO on January 6th.

**Reply by sandriiy** [reply]
I've been trying to figure out what's wrong with my custom permissions for a few days now, it's good that I came across your message, in my orgs `decomposePermissionSetBeta2` causes the same problem, `decomposePermissionSetBeta` works fine though.

**Reply by mcarvin8** [reply]
New issue created here - https://github.com/forcedotcom/cli/issues/3165

**Comment by klouf-threshold** [comment]
Is there an estimated timeframe for when this will be promoted to GA? While I'm certainly glad it's gone through a seemingly successful beta period and had adjustments and fixes made before going GA, it feels like this beta functionality has been forgotten/accepted as GA without actually being promoted as such. The original announcement was over 18 months ago (https://github.com/forcedotcom/cli/blob/main/releasenotes/README.md#2368-april-10-2024)...

Thanks!
